### PR TITLE
Adds additional CFN parameters

### PR DESCRIPTION
**Pull request by userhas404d**
AWS::EC2::Instance

 The following parameters were added:

  - DisableApiTermination
  - EbsOptimized

  The following options were enabled by default:

  - Monitoring

AWS::AutoScaling::AutoScalingGroup

 The following parameters were added:

  - MetricsCollection
  - NotificationConfiguration
  - TerminationPolicies

AWS::AutoScaling::LaunchConfiguration

 The following parameters were added:

  - EbsOptimized

### CODE REVIEW

**Comment by userhas404d** [comment]
go codebuild go

**Comment by userhas404d** [comment]
go codebuild go

**Comment by userhas404d** [comment]
go codebuild go

**Comment by lorengordon** [comment]
> One of the parameters you're exposing is in all caps, which deviates from the way the rest of the parameters are listed. There are two other ASG parameters where the s and g should be lower case

Yes, please. Even acronyms, only capitalize the first letter for parameter/resource names...

File: modules/lx-autoscale/watchmaker-lx-autoscale.template.cfn.yaml [hunk 1]
```diff
@@ -303,6 +330,26 @@ Parameters:
     Default: gp2
     Description: Type of EBS volume to create. Ignored if "AppVolumeDevice" is blank
     Type: String
+  AsgMetrics:
+    Default: ''
+    Description: >-
+      (Optional) The list of Asg metrics to collect. Must define MetricsCollectionGranularity to enable.
+      Define MetricsCollectionGranularity and leave this option blank to collect all metrics
+    Type: CommaDelimitedList
+  AsgNotificaitonTypes:
```
Comment (moskey71) [inline modules/lx-autoscale/watchmaker-lx-autoscale.template.cfn.yaml]: spelling error

File: modules/lx-autoscale/watchmaker-lx-autoscale.template.cfn.yaml [hunk 2]
```diff
@@ -303,6 +330,26 @@ Parameters:
...
+    Default: ''
+    Description: >-
+      (Optional) A list of event types that trigger a notification. Event types can include any of the following
+      types: autoscaling:EC2_INSTANCE_LAUNCH, autoscaling:EC2_INSTANCE_LAUNCH_ERROR, autoscaling:EC2_INSTANCE_TERMINATE,
+      autoscaling:EC2_INSTANCE_TERMINATE_ERROR, and autoscaling:TEST_NOTIFICATION.
+      NOTE: Must be defined in conjunction with AsgSnsArn.
+    Type: CommaDelimitedList
+  AsgSnsA:
```
Comment (moskey71) [inline modules/lx-autoscale/watchmaker-lx-autoscale.template.cfn.yaml]: looks like you lost 'rn' here

**Comment by userhas404d** [comment]
go codebuild go

**Comment by userhas404d** [comment]
go codebuild go

File: modules/lx-autoscale/watchmaker-lx-autoscale.template.cfn.yaml (hunk 1 above)
Comment (userhas404d) [reply modules/lx-autoscale/watchmaker-lx-autoscale.template.cfn.yaml]: fixed

File: modules/lx-autoscale/watchmaker-lx-autoscale.template.cfn.yaml (hunk 2 above)
Comment (userhas404d) [reply modules/lx-autoscale/watchmaker-lx-autoscale.template.cfn.yaml]: updated

File: modules/lx-autoscale/variables.tf [hunk 3]
```diff
@@ -295,3 +295,33 @@ variable "ToggleNewInstances" {
...
+
+variable "InstanceTerminationPolicies" {
+  type        = "string"
+  description = "(Optional) A policy or a list of policies that are used to select the instances to terminate. The policies are executed in the order that you list them."
+  default     = ""
+}
+
+variable "EbsOptimized" {
+  type        = "string"
+  description = "(Optional) Specifies whether the launch configuration is optimized for EBS I/O. This optimization provides dedicated throughput to Amazon EBS and an optimized configuration stack to provide optimal EBS I/O performance"
```
Comment (lorengordon) [inline modules/lx-autoscale/variables.tf]:
```suggestion
  description = "Specifies whether the launch configuration is optimized for EBS I/O. This optimization provides dedicated throughput to Amazon EBS and an optimized configuration stack to provide optimal EBS I/O performance"
```

File: modules/lx-instance/variables.tf [hunk 4]
```diff
@@ -259,3 +259,15 @@ variable "ToggleCfnInitUpdate" {
...
+
+variable "DisableApiTermination" {
+  type        = "string"
+  description = "(Optional) Specifies whether the instance can be terminated through the API."
+  default     = "false"
+}
+
+variable "EbsOptimized" {
+  type        = "string"
+  description = "(Optional) Specifies whether the instance is optimized for Amazon Elastic Block Store I/O. See complete list of supported instances here: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSOptimized.html#ebs-optimization-support"
```
Comment (lorengordon) [inline modules/lx-instance/variables.tf]:
```suggestion
  description = "Specifies whether the instance is optimized for Amazon Elastic Block Store I/O. See complete list of supported instances here: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSOptimized.html#ebs-optimization-support"
```

File: modules/lx-instance/watchmaker-lx-instance.template.cfn.yaml [hunk 5]
```diff
@@ -295,6 +303,22 @@ Parameters:
     Description: >-
       (Optional) List of application log file paths to send to CloudWatch. Example: path01/file01.log, path02/file02.log
     Type: CommaDelimitedList
+  DisableApiTermination:
+    AllowedValues:
+      - 'false'
+      - 'true'
+    Default: 'false'
+    Description: >-
+      (Optional) Specifies whether the instance can be terminated through the API.
```
Comment (lorengordon) [inline modules/lx-instance/watchmaker-lx-instance.template.cfn.yaml]:
```suggestion
      Specifies whether the instance can be terminated through the API.
```

File: modules/lx-instance/watchmaker-lx-instance.template.cfn.yaml [hunk 6]
```diff
@@ -295,6 +303,22 @@ Parameters:
...
+    Type: String
+  EbsOptimized:
+    Default: 'false'
+    Description: >-
+      (Optional) Specifies whether the instance is optimized for Amazon Elastic Block Store I/O.
```
Comment (lorengordon) [inline modules/lx-instance/watchmaker-lx-instance.template.cfn.yaml]:
```suggestion
      Specifies whether the instance is optimized for Amazon Elastic Block Store I/O.
```

File: modules/lx-instance/watchmaker-lx-instance.template.cfn.yaml [hunk 7]
```diff
@@ -746,13 +770,22 @@ Resources:
               VolumeSize: !Ref AppVolumeSize
               VolumeType: !Ref AppVolumeType
           - !Ref 'AWS::NoValue'
+      DisableApiTermination: !If
```
Comment (lorengordon) [inline modules/lx-instance/watchmaker-lx-instance.template.cfn.yaml]: Same points as EbsOptimized on lx-autoscale

File: modules/lx-instance/watchmaker-lx-instance.template.cfn.yaml [hunk 8]
```diff
@@ -746,13 +770,22 @@ Resources:
...
+        - UseApiTermination
+        - true
+        - false
+      EbsOptimized: !If
```
Comment (lorengordon) [inline modules/lx-instance/watchmaker-lx-instance.template.cfn.yaml]: Same points as EbsOptimized on lx-autoscale

File: modules/win-autoscale/variables.tf [hunk 9]
```diff
@@ -264,3 +264,33 @@ variable "ToggleNewInstances" {
...
+
+variable "InstanceTerminationPolicies" {
+  type        = "string"
+  description = "(Optional) A policy or a list of policies that are used to select the instances to terminate. The policies are executed in the order that you list them."
+  default     = ""
+}
+
+variable "EbsOptimized" {
+  type        = "string"
+  description = "(Optional) Specifies whether the launch configuration is optimized for EBS I/O. This optimization provides dedicated throughput to Amazon EBS and an optimized configuration stack to provide optimal EBS I/O performance"
```
Comment (lorengordon) [inline modules/win-autoscale/variables.tf]:
```suggestion
  description = "Specifies whether the launch configuration is optimized for EBS I/O. This optimization provides dedicated throughput to Amazon EBS and an optimized configuration stack to provide optimal EBS I/O performance"
```

File: modules/win-autoscale/watchmaker-win-autoscale.template.cfn.yaml [hunk 10]
```diff
@@ -240,10 +287,25 @@ Parameters:
     Default: 1
     Description: Desired number of instances in the Autoscaling Group
     Type: Number
+  EbsOptimized:
+    AllowedValues:
+      - 'true'
+      - 'false'
+    Default: 'false'
+    Description: >-
+      (Optional) Specifies whether the launch configuration is optimized for EBS I/O. This optimization provides
```
Comment (lorengordon) [inline modules/win-autoscale/watchmaker-win-autoscale.template.cfn.yaml]:
```suggestion
      Specifies whether the launch configuration is optimized for EBS I/O. This optimization provides
```

File: modules/win-autoscale/watchmaker-win-autoscale.template.cfn.yaml [hunk 11]
```diff
@@ -825,6 +905,10 @@ Resources:
               VolumeSize: !Ref AppVolumeSize
               VolumeType: !Ref AppVolumeType
           - !Ref 'AWS::NoValue'
+      EbsOptimized: !If
```
Comment (lorengordon) [inline modules/win-autoscale/watchmaker-win-autoscale.template.cfn.yaml]: Same points as EbsOptimized on lx-autoscale

File: modules/win-instance/variables.tf [hunk 12]
```diff
@@ -236,3 +236,15 @@ variable "ToggleCfnInitUpdate" {
   description = "A/B toggle that forces a change to instance metadata, triggering the cfn-init update sequence"
   default     = "A"
 }
+
+variable "DisableApiTermination" {
+  type        = "string"
+  description = "(Optional) Specifies whether the instance can be terminated through the API."
```
Comment (lorengordon) [inline modules/win-instance/variables.tf]:
```suggestion
  description = "Specifies whether the instance can be terminated through the API."
```

File: modules/win-instance/variables.tf [hunk 13]
```diff
@@ -236,3 +236,15 @@ variable "ToggleCfnInitUpdate" {
...
+  default     = "false"
+}
+
+variable "EbsOptimized" {
+  type        = "string"
+  description = "(Optional) Specifies whether the instance is optimized for Amazon Elastic Block Store I/O. See complete list of supported instances here: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSOptimized.html#ebs-optimization-support"
```
Comment (lorengordon) [inline modules/win-instance/variables.tf]:
```suggestion
  description = "Specifies whether the instance is optimized for Amazon Elastic Block Store I/O. See complete list of supported instances here: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSOptimized.html#ebs-optimization-support"
```

File: modules/win-instance/watchmaker-win-instance.template.cfn.yaml [hunk 14]
```diff
@@ -201,6 +209,22 @@ Parameters:
     Description: >-
       (Optional) List of application log file paths to send to CloudWatch. Example: C:\\dir1\\file1,C:\\dir2\file2,C:\\dir3\\file3
     Type: CommaDelimitedList
+  DisableApiTermination:
+    AllowedValues:
+      - 'false'
+      - 'true'
+    Default: 'false'
+    Description: >-
+      (Optional) Specifies whether the instance can be terminated through the API.
```
Comment (lorengordon) [inline modules/win-instance/watchmaker-win-instance.template.cfn.yaml]:
```suggestion
      Specifies whether the instance can be terminated through the API.
```

File: modules/win-instance/watchmaker-win-instance.template.cfn.yaml [hunk 15]
```diff
@@ -201,6 +209,22 @@ Parameters:
...
+    Type: String
+  EbsOptimized:
+    Default: 'false'
+    Description: >-
+      (Optional) Specifies whether the instance is optimized for Amazon Elastic Block Store I/O.
```
Comment (lorengordon) [inline modules/win-instance/watchmaker-win-instance.template.cfn.yaml]:
```suggestion
      Specifies whether the instance is optimized for Amazon Elastic Block Store I/O.
```

File: modules/win-instance/watchmaker-win-instance.template.cfn.yaml [hunk 16]
```diff
@@ -711,13 +735,22 @@ Resources:
               VolumeSize: !Ref AppVolumeSize
               VolumeType: !Ref AppVolumeType
           - !Ref 'AWS::NoValue'
+      DisableApiTermination: !If
```
Comment (lorengordon) [inline modules/win-instance/watchmaker-win-instance.template.cfn.yaml]: Same points as EbsOptimized on lx-autoscale

File: modules/win-instance/watchmaker-win-instance.template.cfn.yaml [hunk 17]
```diff
@@ -711,13 +735,22 @@ Resources:
...
+        - UseApiTermination
+        - true
+        - false
+      EbsOptimized: !If
```
Comment (lorengordon) [inline modules/win-instance/watchmaker-win-instance.template.cfn.yaml]: Same points as EbsOptimized on lx-autoscale

File: modules/win-autoscale/watchmaker-win-autoscale.template.cfn.yaml [hunk 18]
```diff
@@ -240,10 +287,25 @@ Parameters:
...
+      dedicated throughput to Amazon EBS and an optimized configuration stack to provide optimal EBS I/O performance
+    Type: String
   InstanceRole:
     Default: ''
     Description: (Optional) IAM instance role to apply to the instance(s)
     Type: String
+  InstanceTerminationPolicies:
+    Default: ''
+    Description: >-
+      (Optional) A policy or a list of policies that are used to select the instances to terminate.
```
Comment (lorengordon) [inline modules/win-autoscale/watchmaker-win-autoscale.template.cfn.yaml]: Even if providing just one, it's a list with a single item...

```suggestion
      (Optional) A list of policies that are used to select the instances to terminate.
```

File: modules/win-autoscale/variables.tf [hunk 19]
```diff
@@ -264,3 +264,33 @@ variable "ToggleNewInstances" {
...
+
+variable "AsgNotificationTypes" {
+  type        = "string"
+  description = "(Optional) A list of event types that trigger a notification. Event types can include any of the following types: autoscaling:EC2_INSTANCE_LAUNCH, autoscaling:EC2_INSTANCE_LAUNCH_ERROR, autoscaling:EC2_INSTANCE_TERMINATE, autoscaling:EC2_INSTANCE_TERMINATE_ERROR, and autoscaling:TEST_NOTIFICATION. NOTE: Must be defined in conjunction with ASGSNSARN."
+  default     = ""
+}
+
+variable "InstanceTerminationPolicies" {
+  type        = "string"
+  description = "(Optional) A policy or a list of policies that are used to select the instances to terminate. The policies are executed in the order that you list them."
```
Comment (lorengordon) [inline modules/win-autoscale/variables.tf]:
```suggestion
  description = "(Optional) list of policies that are used to select the instances to terminate. The policies are executed in the order that you list them."
```

File: modules/win-autoscale/variables.tf [hunk 20]
```diff
@@ -264,3 +264,33 @@ variable "ToggleNewInstances" {
...
+}
+
+variable "AsgNotificationTypes" {
+  type        = "string"
+  description = "(Optional) A list of event types that trigger a notification. Event types can include any of the following types: autoscaling:EC2_INSTANCE_LAUNCH, autoscaling:EC2_INSTANCE_LAUNCH_ERROR, autoscaling:EC2_INSTANCE_TERMINATE, autoscaling:EC2_INSTANCE_TERMINATE_ERROR, and autoscaling:TEST_NOTIFICATION. NOTE: Must be defined in conjunction with ASGSNSARN."
+  default     = ""
+}
+
+variable "InstanceTerminationPolicies" {
+  type        = "string"
```
Comment (lorengordon) [inline modules/win-autoscale/variables.tf]: Use native types, join them in the tf module.

```suggestion
  type        = "list"
```

File: modules/win-autoscale/variables.tf [hunk 21]
```diff
@@ -264,3 +264,33 @@ variable "ToggleNewInstances" {
...
+}
+
+variable "AsgSnsArn" {
+  type        = "string"
+  description = "(Optional) The Amazon Resource Name (ARN) of the Amazon Simple Notification Service (SNS) topic to send ASG events to. NOTE: Must be defined in conjunction with ASGNotificationTypes."
+  default     = ""
+}
+
+variable "AsgNotificationTypes" {
+  type        = "string"
```
Comment (lorengordon) [inline modules/win-autoscale/variables.tf]:
```suggestion
  type        = "list"
```

File: modules/win-autoscale/variables.tf [hunk 22]
```diff
@@ -264,3 +264,33 @@ variable "ToggleNewInstances" {
   description = "A/B toggle that forces a change to instance userdata, triggering new instances via the Autoscale update policy"
   default     = "A"
 }
+
+variable "AsgMetrics" {
+  type        = "string"
```
Comment (lorengordon) [inline modules/win-autoscale/variables.tf]:
```suggestion
  type        = "list"
```

File: modules/lx-autoscale/variables.tf [hunk 23]
```diff
@@ -295,3 +295,33 @@ variable "ToggleNewInstances" {
   description = "A/B toggle that forces a change to instance userdata, triggering new instances via the Autoscale update policy"
   default     = "A"
 }
+
+variable "AsgMetrics" {
+  type        = "string"
```
Comment (lorengordon) [inline modules/lx-autoscale/variables.tf]:
```suggestion
  type        = "list"
```

File: modules/lx-autoscale/variables.tf [hunk 24]
```diff
@@ -295,3 +295,33 @@ variable "ToggleNewInstances" {
...
+}
+
+variable "AsgSnsArn" {
+  type        = "string"
+  description = "(Optional) The Amazon Resource Name (ARN) of the Amazon Simple Notification Service (SNS) topic to send ASG events to. NOTE: Must be defined in conjunction with ASGNotificationTypes."
+  default     = ""
+}
+
+variable "AsgNotificationTypes" {
+  type        = "string"
```
Comment (lorengordon) [inline modules/lx-autoscale/variables.tf]:
```suggestion
  type        = "list"
```

File: modules/lx-autoscale/variables.tf [hunk 25]
```diff
@@ -295,3 +295,33 @@ variable "ToggleNewInstances" {
...
+  description = "(Optional) A list of event types that trigger a notification. Event types can include any of the following types: autoscaling:EC2_INSTANCE_LAUNCH, autoscaling:EC2_INSTANCE_LAUNCH_ERROR, autoscaling:EC2_INSTANCE_TERMINATE, autoscaling:EC2_INSTANCE_TERMINATE_ERROR, and autoscaling:TEST_NOTIFICATION. NOTE: Must be defined in conjunction with ASGSNSARN."
+  default     = ""
+}
+
+variable "InstanceTerminationPolicies" {
+  type        = "string"
```
Comment (lorengordon) [inline modules/lx-autoscale/variables.tf]:
```suggestion
  type        = "list"
```

File: modules/lx-autoscale/variables.tf [hunk 26]
```diff
@@ -295,3 +295,33 @@ variable "ToggleNewInstances" {
...
+  description = "(Optional) A policy or a list of policies that are used to select the instances to terminate. The policies are executed in the order that you list them."
```
Comment (lorengordon) [inline modules/lx-autoscale/variables.tf]:
```suggestion
  description = "(Optional) A list of policies that are used to select the instances to terminate. The policies are executed in the order that you list them."
```

**Comment by userhas404d** [comment]
go codebuild go

**Comment by userhas404d** [comment]
go codebuild go

**Pull request merged**